/**
 * @file dialogedittransition.h
 * @author Eliška Křeménková (xkremee00)
 * @brief Declaration of the DialogEditTransition class that handles editing transition timeout and input conditions
 * @date 2025-05-10
 */

#ifndef DIALOGEDITTRANSITION_H
#define DIALOGEDITTRANSITION_H

#include <QDialog>

namespace Ui {
class DialogEditTransition;
}

/**
 * @class DialogEditTransition
 * @brief A dialog window for editing transition.
 *
 * This class provides a simple UI for entering
 * - list of input conditions
 * - timeout
 *
 * It is used as a modal dialog and returns user input
 * through getter methods after acceptance.
 */
class DialogEditTransition : public QDialog
{
    Q_OBJECT

public:
    /**
     * @brief Construct a new Dialog Edit Transition object
     * @param parent Parent widget, default is nullptr
     */
    explicit DialogEditTransition(QWidget *parent = 0);

    /**
     * @brief Destroy the Dialog Edit Transition object
     */
    ~DialogEditTransition();

    /**
     * @brief Set available input pointers to the dialog's combo box
     * @param pointers List of available pointers
     */
    void setInputPointers(const QStringList& pointers);

    /**
     * @brief Get the input conditions from the dialog
     * @return list of conditions stored as string pairs
     */
    QList<QPair<QString, QString> > getInputConditions() const;

    /**
     * @brief Get the timeout of the transition from the dialog
     * @return timeout in milliseconds
     */
    int getTimeout() const;

    /**
     * @brief Set available input condition
     * @param conditions List of available input conditions
     */
    void setInputConditions(const QList<QPair<QString, QString>> &conditions);

    /**
     * @brief Set the timeout of the transition to the dialog's spin box
     * @param timeout Timeout in milliseconds
     */
    void setTimeout(int timeout);

private slots:

    /**
     * @brief Add a new input condition to the list of input conditions inside the dialog
     *
     * Triggered when the user clicks the "Add" button
     * and adds data from input fields (e.g., input symbol and pointer) to the list
     */
    void on_pushButton_add_condition_clicked();

    /**
     * @brief Remove selected input condition from the list of input conditions inside the dialog
     *
     * Triggered when the user clicks the "Remove" button
     * and removes selected condition from the list
     */
    void on_pushButton_remove_condition_clicked();

private:
    /**
     * @brief Pointer to the UI elements defined in the dialogedittransition.ui file.
     *
     * This is automatically generated by Qt's user interface compiler (uic) and provides
     * access to all widgets and layout elements created in the Qt Designer for DialogEditTransition.
     */
    Ui::DialogEditTransition *ui;
};

#endif // DIALOGEDITTRANSITION_H
